[DOXIA-616]

bertysentry opened the pull request:
Broad update to **doxia-module-markdown**

* Properly exposed language class in fenced code blocks
* Re-implemented metadata processing
* Added unit tests
* Added integration tests (with maven-site-plugin)
* Fixed *XhtmlBaseParser* and *XhtmlBaseSink* to properly expose attributes of the PRE and CODE elements
* Fixed *AbstractModuleTest* to use UTF-8 for input files
* Bonus: Fixed DOXIA-542

bertysentry commented on doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownParser.java:
```diff
@@ -24,27 +24,33 @@
+import com.vladsch.flexmark.ext.wikilink.WikiLinkExtension;
+import com.vladsch.flexmark.ext.gfm.strikethrough.StrikethroughExtension;
+
+import org.apache.commons.io.input.CharSequenceReader;
```
I took the liberty to include Commons IO to convert a *StringBuilder* to a *Reader* without going through a *String*. Let me know if that is overkill and you prefer to not to use Commons IO. Thanks!

michael-o replied:
I am fine with that. WE love Apache dogfood.

slachiewicz commented on doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownParser.java:
```diff
@@ -74,48 +80,103 @@
+    {
+        MutableDataSet flexmarkOptions = new MutableDataSet();
+
+        // Emulate Pegdown's behavior
```
What is yours opinion about using this emulation? Should we stay with this or switch to some different profile?

bertysentry replied:
Good question.

Flexmark's default behavior is [CommonMark 0.28](https://spec.commonmark.org/0.28/).

Unit tests weren't covering edge cases like Pegdown vs CommonMark, so I was concerned to break people's documentation and I kept the Pegdown profile. It's really a question of balance between risk of breaking existing stuff, and moving forward with something cleaner and properly specified.

I'm personally in favor of moving forward with CommonMark (just like Web developers are happy that we finally got rid of Microsoft Internet Explorer... :-D)

slachiewicz replied:
Let's try after this PR merge.

michael-o replied:
i would favor move to CommonMark and move master to 1.10.

michael-o replied:
Can we move to CommonMark instead of sticking to ages?

bertysentry replied:
Okay, you're right! Guys, fasten your seatbelts, let's do this! :-)

bertysentry replied:
I'm removing the Pegdown emulation profile option, but I will keep all of the extensions that used to be enabled. We don't want to lose features.

Also, I'm thinking of enabling the [YAML Extension](https://github.com/vsch/flexmark-java/wiki/Extensions#yaml-front-matter), which is very similar to what we're doing to process metadata, but that will be for another issue.

michael-o commented:
Massive...will take a day or two

bertysentry commented:
@michael-o:
> Massive...will take a day or two

 Yeah, sorry for the massive PR. One thing lead to another...

slachiewicz commented:
I'm ok with this changes to be merged. We can test it with our maven website and compare differences.

michael-o commented on doxia-core/src/main/java/org/apache/maven/doxia/parser/XhtmlBaseParser.java:
```diff
@@ -509,7 +509,8 @@ else if ( ( parser.getName().equals( HtmlMarkup.CODE.toString() ) )
                 || ( parser.getName().equals( HtmlMarkup.TT.toString() ) ) )
         {
-            sink.inline( SinkEventAttributeSet.Semantics.CODE );
+            attribs.addAttributes( SinkEventAttributeSet.Semantics.CODE );
```
Interesting, is this a semantical change?

bertysentry replied:
Before this change, an *inline* event was added with the *CODE* semantic, and getting rid of all potential attributes of the element being parsed.

After this change, we add the *inline* event, with the *CODE* semantic **in addition** to the attributes of the element being parsed.

So, `<code class="language-java">` becomes an *inline* event, with `Semantics = "code"` and `Class = "language-java"`.

Note this reflects what was already done in **Xhtml5BaseParser.java**. I think it was an omission.

michael-o commented on doxia-core/src/main/java/org/apache/maven/doxia/sink/impl/Xhtml5BaseSink.java:
```diff
@@ -1905,7 +1905,9 @@ private void inlineSemantics( SinkEventAttributes attributes, String semantic,
         {
-            writeStartTag( tag );
+            SinkEventAttributes attributesNoSemantics = ( SinkEventAttributes ) attributes.copyAttributes();
+            attributesNoSemantics.removeAttribute( SinkEventAttributes.SEMANTICS );
```
Why is this necessary?

bertysentry replied:
*inline* Sink events may have attributes that we want to leverage (notably the *Class* attribute).

Example: *inline* Sink event with `Semantics = "code"` and `Class = "language-java"` will be output as `<code class="language-java">`.

Note: We don't want to output `<code semantics="code" class="language-java">` and that's why we need to remove the *SEMANTICS* attribute.

Admittedly, the *inlineSemantics()* function is not the most elegant piece of code but I didn't want to reshuffle the XHTML Sink too much, as it's used in many different places.

michael-o commented on doxia-modules/doxia-module-markdown/pom.xml:
```diff
@@ -72,5 +72,80 @@ under the License.
+
+      <!-- install -->
+      <plugin>
+        <artifactId>maven-install-plugin</artifactId>
```
Aren't they available when `verify` run because this is run after `install`?

bertysentry replied:
I wanted to make sure `mvn verify` runs successfully on the reactor, without the need to run `mvn install` before.

Also, the artifacts produced by other modules are not retrieved by **maven-invoker-plugin** (even with its `<extraArtifacts>` configuration property). That would be a nice enhancement on this plugin! ;-)

michael-o replied:
hm... I need to think about this. It is somewhat hacky, but I do understand your intentions.

bertysentry replied:
It is hacky, I agree, especially as it specifies other artifacts by their path relative to this module. At the same time, it's really the purpose of **maven-install-plugin**: install artifacts in a Maven repository, which is what we're doing here so that **maven-invoker-plugin** has all it needs to run integration tests.

I'm open to alternatives (as that would be useful for other multi-module projects!).

michael-o replied:
@rfscholte @khmarbaise Do you have an opinion here?

michael-o commented on doxia-modules/doxia-module-markdown/src/it/settings.xml:
```diff
@@ -0,0 +1,55 @@
+      </activation>
+      <repositories>
+        <repository>
+          <id>1local.central</id>
```
Is the preceding `1` intentional?

bertysentry replied:
Ha! Good question. To be honest, I copy-pasted this **maven-invoker-plugin** configuration from another project. I remember struggling with the order of repositories and someone suggested prefixing the repository with "1" so that it comes before "all". But I'm not quite sure...

If you recommend to remove "1", let me know.

michael-o replied:
Leave as-is.

michael-o commented on doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownParser.java:
```diff
@@ -74,48 +80,103 @@
+        ) );
+
+        // Disable wrong apostrophe replacement
+        flexmarkOptions.set( TypographicExtension.SINGLE_QUOTE_UNMATCHED, "&apos;" );
```
Does this address DOXIA-542?

bertysentry replied:
Indeed, it does! :-)

michael-o replied:
I'd prefer this to be separate for documentation reasons.

bertysentry replied:
Understood!

michael-o replied:
Let's finish this first and then put 542 on top.

michael-o commented on doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownParser.java:
```diff
@@ -130,133 +191,98 @@ public void parse( Reader source, Sink sink )
-                    html.append( StringEscapeUtils.escapeXml( value ) );
-                    html.append( "\' />" );
+                    html.append( "<meta name='" );
+                    html.append( HtmlTools.escapeHTML( key, true ) );
```
Is this is a superset of `escapeXml`?

bertysentry replied:
*escapeXml()* is deprecated (as well as the entire *StringEscapeUtils* class), that's why I replaced it with its equivalent in this very project. It's supposed to do exactly the same.

michael-o replied:
I understand that, but the Sink says XHTML, so we need at least XML escapes. Just checked the code, all is fine. It runs in XML mode. You can drop the `true`. This is the default value.

michael-o commented on doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownParser.java:
```diff
@@ -130,133 +191,98 @@ public void parse( Reader source, Sink sink )
-            while ( firstNode != null && !( firstNode instanceof Heading ) )
+            // Skip the comment nodes
+            Node firstNode = documentRoot.getFirstChild();
+            while ( firstNode != null && firstNode instanceof HtmlCommentBlock )
```
Thid does not retain HTML style commments?

bertysentry replied:
This section is about retrieving the a potential title in the document. If the first meaningful (non-comment) node is a *Heading*, then it is considered as the title. Here, we're skipping comment nodes to check the first non-comment node.

Comment nodes are preserved in the document that is produced.

michael-o replied:
OK.

bertysentry commented:
@michael-o I removed code in reference to DOXIA-542 (quote vs apostrophe) and associated tests. I removed the `true` flag in *HtmlTools.escapeHTML()* calls as suggested. Please review the updated code. Thank you!

bertysentry commented:
Guys, I updated the PR with all suggested changes. Is there anything else I can do? I would like to have this merged so I can submit my PR on DOXIA-542 (apostrophes vs quotes). Thanks!

michael-o commented:
OK, let me go through again and test it.

michael-o commented on doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownParser.java:
```diff
@@ -130,133 +184,98 @@ public void parse( Reader source, Sink sink )
                     haveTitle = true;
                     html.append( "<title>" );
-                    html.append( StringEscapeUtils.escapeXml( value ) );
+                    html.append( HtmlTools.escapeHTML( value, false ) );
```
Why is here a `false` and not just one arg like the rest of the calls?

bertysentry replied:
The only difference between *false* and *true* in *HtmlTools.escapeHTML()* is the handling of the apostrophe (should it be replaced with `&apos;` or not). As we're inside a `<title>` element, there is no need to replace the apostrophe.

However, on line 218 and 221, we're inside the attribute of an element, i.e. inside quotes, and that's why we need to escape the apostrophe here.

Note that the presence (and absence) of apostrophe in the title and metadatas is covered with unit tests (see *testMetadataSinkEvent()* in **MarkdownParserTest.java**)

michael-o replied:
I am not really happy with that because we have an XHTML (HTML) sink and `false` means that it uses HTML entities which XML does not have.

bertysentry replied:
Happy New Year! 😉

I understand your concern, but I think that the parameter names are misleading.

More importantly, the *XhtmlBaseSink* **itself** uses `HtmlTools.escapeHTML(text, false)`:

* For example, all XHTML content is inserted with [XhtmlBaseSink.content()](https://github.com/apache/maven-doxia/blob/74c3e876b5c7a044f5a6b7ad7be6ea3ac10a6f49/doxia-core/src/main/java/org/apache/maven/doxia/sink/impl/XhtmlBaseSink.java#L2176)
* *XhtmlBaseSink.content()* calls [XhtmlBaseSink.escapeHTML()](https://github.com/apache/maven-doxia/blob/74c3e876b5c7a044f5a6b7ad7be6ea3ac10a6f49/doxia-core/src/main/java/org/apache/maven/doxia/sink/impl/XhtmlBaseSink.java#L2201)

and *XhtmlBaseSink.escapeHTML()* is simply:

```java
    /**
     * Forward to HtmlTools.escapeHTML( text ).
     *
     * @param text the String to escape, may be null
     * @return the text escaped, "" if null String input
     * @see org.apache.maven.doxia.util.HtmlTools#escapeHTML(String)
     */
    protected static String escapeHTML( String text )
    {
        return HtmlTools.escapeHTML( text, false );
    }
```

So, calling `HtmlTools.escapeHTML( text, false )` is really the only proper way to produce XHTML content that will be parsed with the XhtmlSink parser. And it is consistent with Doxia's own XhtmlSink.

michael-o replied:
I am sick and tired of the codebase we have. Hacks after hacks with escaping rules w/o any consistency. BUT you are right, lets keep it consistent even if it is wrong.

bertysentry replied:
I hear you! I guess *escapeHTML()* should be renamed *escapeXHTML()*, as it's what it actually does. And the *xml* param should really be just about apostrophes. Anyway, that's how it is now...

Thank you for the review, @michael-o !

michael-o replied:
Even more than that. HTML5 requires so little escaping because we use UTF-8 now. I see little reason to replace one or two byte characters in UTF-8 with 5 bytes of entities. Insane. This has been done 10-15 years ago, but is obsolete now.

michael-o commented on doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownParser.java:
```diff
@@ -130,133 +184,98 @@ public void parse( Reader source, Sink sink )
                 TextCollectingVisitor collectingVisitor = new TextCollectingVisitor();
                 String headingText = collectingVisitor.collectAndGetText( firstNode );
-                html.append( StringEscapeUtils.escapeXml( headingText ) );
+                html.append( HtmlTools.escapeHTML( headingText, false ) );
```
Same here...

bertysentry replied:
Same answer as above.

michael-o commented on doxia-modules/doxia-module-markdown/src/test/resources/fenced-code-block.md:
````diff
@@ -0,0 +1,5 @@
+Below code is Java:
+
+```java
+System.out.println(helloWorld);
````
Quotes around literal missing

bertysentry replied:
Nah, it's on purpose. This is just a test resource that is translated as Sink events. But the presence of quotes around `helloWorld` breaks the *text* Sink event in several fragment, which is annoying for my test.

But we can consider *helloWorld* is a *String* variable defined somewhere else.

Again, this is just an example for the unit tests. We don't need valid Java code here.

michael-o commented:
I don't have any objection anymore. Let me run the tests and then I will merge.

michael-o commented:
Tested, when I run the general IT one site is missing:
```
192.168.1.52 - - [02/Jan/2021 13:52:48] "GET /general/target/site/quotes.html HTTP/1.1" 404 -
```
```
mosipov@bsd1srv:/usr/home/mosipov/var/Projekte/maven-doxia/doxia-modules/doxia-module-markdown/target/it/general/target/site (DOXIA-616 *)
$ tree
.
├── css
│   ├── maven-base.css
│   ├── maven-theme.css
│   ├── print.css
│   └── site.css
├── images
│   ├── collapsed.gif
│   ├── expanded.gif
│   ├── external.png
│   ├── icon_error_sml.gif
│   ├── icon_info_sml.gif
│   ├── icon_success_sml.gif
│   ├── icon_warning_sml.gif
│   ├── logos
│   │   ├── build-by-maven-black.png
│   │   ├── build-by-maven-white.png
│   │   └── maven-feather.png
│   └── newwindow.png
├── index.html
└── metadata.html
```

bertysentry commented:
> Tested, when I run the general IT one site is missing:
> 
> ```
> 192.168.1.52 - - [02/Jan/2021 13:52:48] "GET /general/target/site/quotes.html HTTP/1.1" 404 -
> ```

Ha! Good catch! I forgot to remove the reference to **quotes.html** in **site.xml**, that's all. I'll fix this. Thanks!

bertysentry commented:
Reference to **quotes.html** has been removed. You shouldn't get a 404 for this now. Thank you!

michael-o commented:
The changeset seems to be fine. There is now expected fallout in MSITE:
```
[INFO] Building: doxia-formats/pom.xml
[INFO] run post-build script verify.groovy
[INFO]   The post-build script did not succeed. assert content.contains( '<div class="source"><pre class="prettyprint linenums">code block' )
       |       |
       |       false
       <!DOCTYPE html>
       <!--
        | Generated by Apache Maven Doxia Site Renderer 1.9.2 from src/site/markdown/markdown.md at 2021-01-02
        | Rendered using Apache Maven Fluido Skin 1.8
       -->
       <html xmlns="http://www.w3.org/1999/xhtml" lang="en">
         <head>
           <meta charset="UTF-8" />
           <meta name="viewport" content="width=device-width, initial-scale=1" />
           <meta name="generator" content="Apache Maven Doxia Site Renderer 1.9.2" />
           <title>Doxia formats tests &#x2013; Markdown Format works</title>
           <link rel="stylesheet" href="./css/apache-maven-fluido-1.8.min.css" />
           <link rel="stylesheet" href="./css/site.css" />
           <link rel="stylesheet" href="./css/print.css" media="print" />
           <script src="./js/apache-maven-fluido-1.8.min.js"></script>
         </head>
         <body class="topBarDisabled">
           <div class="container-fluid">
             <header>
               <div id="banner">
                 <div class="pull-left"><div id="bannerLeft"><h2>Doxia formats tests</h2>
       </div>
       </div>
                 <div class="pull-right"></div>
                 <div class="clear"><hr/></div>
               </div>

               <div id="breadcrumbs">
                 <ul class="breadcrumb">
               <li id="publishDate">Last Published: 2021-01-02<span class="divider">|</span>
       </li>
                 <li id="projectVersion">Version: 1.0-SNAPSHOT</li>
                 </ul>
               </div>
             </header>
             <div class="row-fluid">
               <header id="leftColumn" class="span2">
                 <nav class="well sidebar-nav">
         <ul class="nav nav-list">
         </ul>
                 </nav>
                 <div class="well sidebar-nav">
                   <hr />
                   <div id="poweredBy">
                     <div class="clear"></div>
                     <div class="clear"></div>
                     <div class="clear"></div>
       <a href="http://maven.apache.org/" title="Built by Maven" class="poweredBy"><img class="builtBy" alt="Built by Maven" src="./images/logos/maven-feather.png" /></a>
                   </div>
                 </div>
               </header>
               <main id="bodyColumn"  class="span10" >
       <!---
       Licensed to the Apache Software Foundation (ASF) under one
       or more contributor license agreements.  See the NOTICE file
       distributed with this work for additional information
       regarding copyright ownership.  The ASF licenses this file
       to you under the Apache License, Version 2.0 (the
       "License"); you may not use this file except in compliance
       with the License.  You may obtain a copy of the License at

         http://www.apache.org/licenses/LICENSE-2.0

       Unless required by applicable law or agreed to in writing,
       software distributed under the License is distributed on an
       "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
       KIND, either express or implied.  See the License for the
       specific language governing permissions and limitations
       under the License.
       -->
       <section>
       <h2><a name="Markdown_Format_works"></a>Markdown Format works</h2>
       <p>But &#x2018;quotes&#x2019; and &#x201c;double quotes&#x201d; were stripped from HTML result with DOXIA 1.3:
       see <a class="externalLink" href="https://issues.apache.org/jira/browse/DOXIA-473">DOXIA-473</a>.</p>
       <ul>
       <li><a href="#Markdown_Format_works">Markdown Format works</a>
       <ul>
       <li><a href="#Subsection">Subsection</a></li></ul></li></ul>
       <section>
       <h3><a name="Subsection"></a>Subsection</h3>
       <p>Missing <code>monospaced</code> support <a class="externalLink" href="https://issues.apache.org/jira/browse/DOXIA-597">DOXIA-597</a></p>
       <p>Missing code block color <a class="externalLink" href="https://issues.apache.org/jira/browse/DOXIA-571">DOXIA-571</a>:</p>

       <div class="source"><pre class="prettyprint linenums"><code>code block
       Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)
       Maven home: D:\apache-maven-3.6.3\apache-maven\bin\..
       Java version: 1.8.0_232, vendor: AdoptOpenJDK, runtime: C:\Program Files\AdoptOpenJDK\jdk-8.0.232.09-hotspot\jre
       Default locale: en_US, platform encoding: Cp1250
       OS name: &quot;windows 10&quot;, version: &quot;10.0&quot;, arch: &quot;amd64&quot;, family: &quot;windows&quot;
       </code></pre></div></section></section>
               </main>
             </div>
           </div>
           <hr/>
           <footer>
             <div class="container-fluid">
               <div class="row-fluid">
                   <p>Copyright &#169;      2021..</p>
               </div>
             </div>
           </footer>
         </body>
       </html>
[INFO]           doxia-formats/pom.xml ............................ FAILED (5.3 s)
```

Can you have a look and create a followup PR in MSITE?

michael-o commented:
Please also proceed with your other PRs.

bertysentry commented:
Thank you @michael-o , I will look into MSITE tomorrow and submit the PR for DOXIA-542.

jira-importer commented:
Resolve #785
